src/Common/thread_pool/mpmc_bounded_queue.hpp

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2010-2011 Dmitry Vyukov. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided
+// that the following conditions are met:
+//
+//   1. Redistributions of source code must retain the above copyright notice,
+//   this list of
+//      conditions and the following disclaimer.
+//
+//   2. Redistributions in binary form must reproduce the above copyright
+//   notice, this list
+//      of conditions and the following disclaimer in the documentation and/or
+//      other materials
+//      provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY DMITRY VYUKOV "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+// EVENT
+// SHALL DMITRY VYUKOV OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+// OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+// The views and conclusions contained in the software and documentation are
+// those of the authors and
+// should not be interpreted as representing official policies, either expressed
+// or implied, of Dmitry Vyukov.
+
+#pragma once
+
+#include <atomic>
+#include <type_traits>
+#include <vector>
+#include <stdexcept>
+
+namespace tp
+{
+
+/**
+ * @brief The MPMCBoundedQueue class implements bounded
+ * multi-producers/multi-consumers lock-free queue.
```

Comment:
It is difficult to believe that we need anything other than mutex/condvar.
Performance differences should be proven in real-world scenarios.

Reply:
> we've observed cases where creating a new thread is slow and a lot of workers are waiting on mutex and ClickHouse is unable to serve even monitoring requests

It might be actually overloaded, please check `OSCPUWaitMicroseconds` metric and `OSCPUWaitMicroseconds` / `OSCPUVirtualTimeMicroseconds` ratio

Reply:
That was our initial thoughts too. I'll try to collect some short article with the details (too much to write).